batch-result-download.ipynb

```diff
@@ -100,43 +62,36 @@
-    "                \n",
-    "                destinationPointSetName = destinationPointSetId\n",
-    "                if namesFetched:\n",
-    "                    destinationPointSetName = destinations[destinationPointSetId]\n",
-    "                \n",
-    "                gridQuery = urllib.parse.urlencode({'cutoff': cutoff, 'percentile': percentile, 'destinationPointSetId': destinationPointSetId})\n",
-    "                gridUrl = resultUrl + analysisId + '/grid/geotiff?' + gridQuery\n",
-    "                \n",
-    "                # Request a signed S3 url from the Conveyal API\n",
-    "                r = requests.get(gridUrl, headers = headers, verify = False)\n",
+    "    print('Processing ' + analysisId)\n",
+    "    status = 202;\n",
     "    \n",
-    "                if r.status_code == 403:\n",
-    "                    print('Unauthorized access. Your authorization token may be invalid or expired.')\n",
+    "    while (status == 202):\n",
+    "        resultUrl = baseUrl + analysisId + '/all'\n",
+    "        # Request a signed S3 url from the Conveyal API\n",
+    "        r = requests.get(resultUrl, headers = headers, verify = False)\n",
+    "        if r.status_code == 403:\n",
+    "            print('Unauthorized access. Your authorization token may be invalid or expired.')\n",
+    "            \n",
+    "        elif r.status_code == 404:\n",
+    "            print('Results not found. Check your analysisId values')\n",
     "        \n",
-    "                elif r.status_code != 200:\n",
-    "                    print('Error: ' + r.text)\n",
+    "        elif r.status_code == 202:\n",
+    "            print(r.text + ' Retrying in 15 seconds.')\n",
+    "            time.sleep(15)\n",
     "        \n",
-    "                else:\n",
-    "                    # From the signed S3 url, fetch the grid as a geotiff\n",
-    "                    grid = requests.get(r.json()['url'], verify = False)\n",
-    "                    # Save response from Conveyal Analysis to a local .geotiff file\n",
-    "                    with open('results/'  + ('-').join((analysisId, destinationPointSetName, str(cutoff) + 'min', str(percentile) + 'pct')) + '.geotiff', 'wb') as f:\n",
-    "                        for chunk in grid.iter_content(chunk_size=128):\n",
-    "                            f.write(chunk)\n",
-    "                        f.close()"
+    "        elif (r.status_code != 200):\n",
+    "            print('Error: ' + r.text)\n",
+    "        \n",
+    "        status = r.status_code\n",
+    "            \n",
+    "    if (status == 200):\n",
+    "        zipRequest = requests.get(r.json()['url'], verify = False)\n",
+    "\n",
+    "        # Save result to a local .zip file\n",
+    "        with open('results/'  + analysisId + '.zip', 'wb') as f:\n",
+    "            for chunk in zipRequest.iter_content(chunk_size=128):\n",
```

Comment:
Although the library's syntax implies this is operating on a stream of chunks (and the documentation at https://fr.python-requests.org/en/latest/api.html also seems to imply this is the point of chunk_size), I wonder if this may actually buffer the entire response in memory. While that wouldn't be much of an issue when working with one geotiff at a time, it's probably not ideal to buffer an entire zip containing an unknown number of geotiffs. Commentary on stackoverflow seems to confirm this concern: https://stackoverflow.com/q/16694907/778449

There are also several edge cases that could occur when downloading and saving a large file, which might be better handled by a purpose-built "save a file" method instead of a simple loop over the content.

Side note: chunk_size is in bytes according to the Requests docs. 128 bytes is a very small chunk size for file IO. It is probably more efficient to work with larger sizes. But ideally a detail like this will be encapsulated in whatever method we use to save the file.

I will research better ways to handle this.

Reply:
Corresponding PR at #3.